### PR TITLE
Support Unicode Directories

### PR DESCRIPTION
**Pull request by EduardoAraujoB**
Solves #169

### CODE REVIEW

File: jscomp/bsb/bsb_ninja_file_groups.ml [hunk 1]
```diff
@@ -124,7 +124,7 @@ let emit_module_build
   let output_ast = rel_proj_dir // (impl_dir // basename output_ast) in
   let output_iast = rel_proj_dir // (intf_dir // basename output_iast) in
   let ast_deps =
-    Format.asprintf "(:ast_deps %s %s)"
+    Format.asprintf {|(:ast_deps "%s" "%s")|}
```
Comment (anmonteiro) [inline jscomp/bsb/bsb_ninja_file_groups.ml:127]: If you use `%S`, OCaml will escape it for us.

```suggestion
    Format.asprintf "(:ast_deps %S %S)"
```

File: jscomp/bsb/bsb_ninja_file_groups.ml [hunk 2]
```diff
@@ -237,13 +237,18 @@ let handle_files_per_dir
...
+    Buffer.add_char buf '"';
     Buffer.add_char buf '\n';
     if group.subdirs <> [] then begin
       Buffer.add_string buf "(dirs :standard";
       Ext_list.iter group.subdirs (fun subdir ->
         Buffer.add_char buf ' ';
-        Buffer.add_string buf subdir);
+        Buffer.add_char buf '"';
+        Buffer.add_string buf subdir;
+        Buffer.add_char buf '"';
```
Comment (anmonteiro) [inline jscomp/bsb/bsb_ninja_file_groups.ml:250]: I just pushed a commit to this branch exposing this function. Would be cool to use it instead.

```suggestion
        Buffer.add_string buf (Ext_filename.maybe_quote_for_dune subdir);
```

File: jscomp/bsb/bsb_ninja_gen.ml [hunk 3]
```diff
@@ -42,7 +42,9 @@ let bsc_lib_includes (bs_dependencies : Bsb_config_types.dependencies) =
 
 let generate_ppxlib_source ~subdir ~ppx_config buf =
   Buffer.add_string buf "(subdir ";
+  Buffer.add_char buf '"';
   Buffer.add_string buf (subdir // Literals.melange_eobjs_dir);
+  Buffer.add_char buf '"';
```
Comment (anmonteiro) [inline jscomp/bsb/bsb_ninja_gen.ml:47]: Same as above 

```suggestion
  Buffer.add_string buf (Ext_filename.maybe_quote_for_dune (subdir // Literals.melange_eobjs_dir));
```

File: jscomp/bsb/bsb_ninja_gen.ml [hunk 4]
```diff
@@ -160,7 +162,9 @@ let output_ninja_and_namespace_map
         ~dir:namespace_dir ns
         bs_file_groups;
       Buffer.add_string buf "(subdir ";
+      Buffer.add_char buf '"';
       Buffer.add_string buf namespace_dir;
+      Buffer.add_char buf '"';
```
Comment (anmonteiro) [inline jscomp/bsb/bsb_ninja_gen.ml:167]: Same as above

```suggestion
      Buffer.add_string buf (Ext_filename.maybe_quote_for_dune namespace_dir);
```

File: jscomp/bsb/bsb_ninja_gen.ml [hunk 5]
```diff
@@ -183,9 +187,13 @@ let output_ninja_and_namespace_map
       Ext_path.rel_normalized_absolute_path ~from:root_dir per_proj_dir
     in
     Buffer.add_string buf "(subdir ";
+    Buffer.add_char buf '"';
     Buffer.add_string buf subd;
+    Buffer.add_char buf '"';
     Buffer.add_string buf "(data_only_dirs ";
+    Buffer.add_char buf '"';
     Buffer.add_string buf Literals.melange_eobjs_dir;
+    Buffer.add_char buf '"';
```
Comment (anmonteiro) [inline jscomp/bsb/bsb_ninja_gen.ml:196]: Also same as above.

```suggestion
    Buffer.add_string buf (Ext_filename.maybe_quote_for_dune subd);
    Buffer.add_string buf "(data_only_dirs ";
    Buffer.add_string buf (Ext_filename.maybe_quote_for_dune Literals.melange_eobjs_dir);
```

File: jscomp/bsb/bsb_ninja_rule.ml [hunk 6]
```diff
@@ -125,7 +125,9 @@ let make_custom_rules
     Buffer.add_string buf " -root ";
     Buffer.add_string buf global_config.per_proj_dir;
     Buffer.add_string buf " -cwd ";
+    Buffer.add_char buf '"';
     Buffer.add_string buf cur_dir;
+    Buffer.add_char buf '"';
```
Comment (anmonteiro) [inline jscomp/bsb/bsb_ninja_rule.ml:130]:
```suggestion
    Buffer.add_string buf (Ext_filename.maybe_quote_for_dune cur_dir);
```

File: jscomp/bsb/bsb_package_specs.ml [hunk 7]
```diff
@@ -135,11 +135,11 @@ let bs_package_output = "-bs-package-output"
 let package_flag ({ format; suffix } : spec) dir =
   Ext_string.inter2
     bs_package_output
-    (Ext_string.concat5
+    (String.make 1 '"' ^ Ext_string.concat5
```
Comment (anmonteiro) [inline jscomp/bsb/bsb_package_specs.ml:138]: we can just call `Ext_filename.maybe_quote_for_dune` here too.

**Comment by EduardoAraujoB** [comment]
thanks, didn't knew about the `Ext_filename.maybe_quote_for_dune`, will apply the fixes asap

**Comment by EduardoAraujoB** [comment]
@anmonteiro there is a problem with the `maybe_quote_for_dune` helper, it's not escaping properly the unicode characters
```
(progn (dynamic-run bsb_helper -root /home/eduardo/Downloads/ocaml-reason/melange-basic-template -cwd "./src/\206\179\206\189\207\137\207\129\206\175\206\182\207\137" %{ast_deps}) (run bsc -I . -I ..    -w a -bs-package-name melange-basic-project -bs-package-output commonjs:"./src/\206\179\206\189\207\137\207\129\206\175\206\182\207\137":.bs.js -o tuturuu.cmj %{inputs})))
```
as you can see, it's escaping the unicode directories to numbers, instead of just the unicode text quoted

**Comment by EduardoAraujoB** [comment]
So, here is the problem, by using the `Format.asprintf "%S"` on the `Ext_filename.maybe_quote_for_dune` OCaml will ecape it to something like this `src/\206\179\206\189\207\137\207\129\206\175\206\182\207\137`, but using the `Format.asprintf {| "%s" |}` it will generate this `src/γνωρίζω` (which is the unicode name of the folder)

**Comment by EduardoAraujoB** [comment]
anyway, I'm still having problems to know what is the correct way for dune to recognize that paths

**Comment by anmonteiro** [comment]
dune supports this directly.

**Pull request closed without merging**